compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFusePackIntoLoop.cpp

```diff
@@ -78,14 +78,28 @@ void AMDAIEFusePackIntoLoopPass::runOnOperation() {
     return;
   }
 
+  if (fusePackDepth < 1) {
+    funcOp->emitOpError("Invalid depth of pack ops for fusion.");
+    return signalPassFailure();
+  }
+
   // Based on the `fusePackDepth`, we would continue to greedily fuse the
   // producer tensor.pack ops.
   LoopLikeOpInterface loops = cast<LoopLikeOpInterface>(scfLoopOp);
   for (unsigned depth = 1; depth <= fusePackDepth; depth++) {
-    // Search the compute op and its producer slices within the For loop.
+    // Search the compute op and its producer slices.
     BlockArgument bbArg = loops.getRegionIterArgs()[0];
     SmallVector<tensor::ExtractSliceOp> sliceOps;
-    for (auto user : bbArg.getUsers()) {
+    SmallVector<Operation *> allUsers(bbArg.getUsers().begin(),
+                                      bbArg.getUsers().end());
+    while (!allUsers.empty()) {
+      auto user = allUsers.pop_back_val();
+      if (isa<tensor::ExtractSliceOp>(user)) {
+        allUsers.insert(allUsers.begin(), user->getUsers().begin(),
+                        user->getUsers().end());
+        continue;
+      }
+
```

Comment:
Please add a comment about the logic.

From what it seems : it's taking up initial users -> adding more users to the beginning of the search party if it bumps into tensor.extract_slice

If we expect only one `linalg.generic` in the body of the for loop - why don't we encash on that?

I'm trying to understand and see if we can get rid of the expensive `allUsers.insert` operation.

Reply:
Thanks @Abhishek-Varma for providing the suggestion! I've simplified the code logic.